### ISSUE TITLE
Unexpected behaviour using difference with Polygons

### ISSUE THREAD

**Issue by Korred** [shapely/shapely#605]
## Expected behavior and actual behavior.

Removing the intersection (difference) of two Polygons from both Polygons should result in two MultiPolygons which consist of two Polygons. However, the actual returned objects are a Polygon and a MultiPolygon. 

The images below show the problem (especially the last one where two "visually" disconnected polygons are treated as one)

**Intersecting Polygons:**
![alt text](https://i.imgur.com/aUewPJ5.png)
![alt text](https://i.imgur.com/85gk8YL.png)

**Intersection:**
![alt text](https://i.imgur.com/OSONg2D.png)

**Polygon (which should be a MultiPolygon):**
![alt text](https://i.imgur.com/cflnIih.png)

Edit1: Enabling edges shows the connection between both polygons:
![alt text](https://i.imgur.com/pAMsAlb.png)

Edit2: Interesting enough, `a.symmetric:_difference(b)` returns a MultiPolygon with 4 Polygons as expected

## Steps to reproduce the problem.

```
from shapely.geometry import Polygon as plg

a = plg([(489.73, 0), (498.96, 0), (662.21, 282.74), (652.97, 282.74)])
b = plg([(494.06, 282.74), (484.83, 282.74), (648.07, 0), (657.31, 0)])

inter = a.intersection(b) # is a Polygon

a_wo_inter = a.difference(inter) # should be a MultiPolygon but its not! :(
b_wo_inter = b.difference(inter) # is a Multipolygon

print("Intersection:", inter)
print("Polygon A without intersection:", a_wo_inter)
print("Polygon B without intersection:", b_wo_inter)
```

## Operating system
Windows 10 Pro (1803) Build: 17134.112
Debian 8.8 Kernel Version: 4.13.0-1011-gcp

## Shapely version and provenance
Installed using conda (also tried using the windows-wheel)
shapely                   1.6.4            py36h2a969d5_0 (Python 3.6.4)

Installed using repl.it [repl-link](https://repl.it/repls/ArtisticPhysicalDebugging)
Shapely-1.6.4.post1-cp36-cp36m-manylinux1_x86_64.whl  (Python 3.6.1)

**Comment by dr-jts** [comment]
This isn't a bug.  It's expected behaviour when computing overlay operations using finite-precision floating point arithmetic.  The results of overlay operations in general will involve some rounding of  coordinates resulting from line segment intersection.  This in turn leads to the situation observed above.

JTS will shortly gain the ability to specify a precision model to be used when computing overlay operations.  This uses snap-rounding to ensure that all constructed coordinates obey the precision model.  Performing the above computation using this new code (with a precision of 1e-2) produces the desired result.

**Comment by Korred** [comment]
@dr-jts Incredible to get an response to something I posted 2 years ago 😮 

Since Shapely is a wrapper around GEOS, which in turn a port of JTS, I will have to wait for both the JTS and GEOS changes in order to use the new code, am I right?

Is there another Python wrapper that uses JTS directly?

**Comment by dr-jts** [comment]
> @dr-jts Incredible to get a response to something I posted 2 years ago 😮

I'm going through all the tickets I can find which report problems with JTS/GEOS overlay, so I can verify how the new code works on them.  So far am batting 100... :)

> Since Shapely is a wrapper around GEOS, which in turn is a port of JTS, I will have to wait for both the JTS and GEOS changes in order to use the new code, am I right?

Correct.  Hopefully the GEOS changes will land by Q3 2020.
 
> Is there another Python wrapper that uses JTS directly?

Not that I'm aware of.

**Comment by Korred** [comment]
@dr-jts Thanks again for the feedback!
Good luck on your endeavours! I'm looking forward to the new code changes!

**Comment by sgillies** [comment]
Thanks @dr-jts !

**Comment by dr-jts** [comment]
Thanks also goes to @pramsey who is working hard on porting the JTS code to GEOS.

And we'll be looking for beta-testers!

**Comment by jorisvandenbossche** [comment]
@Korred another two years, and just to confirm that those changes have landed in GEOS ànd are exposed in Shapely 2.0 (beta version released to test). Using your example from above:

```
>>> a.difference(inter)
<POLYGON ((652.97 282.74, 662.21 282.74, 498.96 0, 489.73 0, 568.9 137.126, ...>

>>> a.difference(inter, grid_size=0.01)
<MULTIPOLYGON (((498.96 0, 489.73 0, 568.9 137.13, 573.52 129.13, 498.96 0))...>
```

(and visually inspecting this resulting MultiPolygon shows that it is now the expected result)

**Comment by Korred** [comment]
Wow! Thanks for getting back! Didn't expect that at all.

**Comment by jorisvandenbossche** [comment]
Given the explanation above, and that this can be solved with shapely 2.0 fixed precision overlays, closing this issue.
